src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala

```diff
@@ -362,19 +374,32 @@ class FastSync(
       }
     }
 
-    private def waitingForBranchResolution: Receive = handleStatus orElse {
+    private def waitingForBranchResolution: Receive = handleStatus orElse handleRequestFailure orElse {
       case FastSyncBranchResolverActor.BranchResolvedSuccessful(firstCommonBlockNumber, newMasterPeer) =>
+        log.debug(
+          s"resolved branch with first common block number $firstCommonBlockNumber for new master peer $newMasterPeer"
+        )
         // Reset the batch failures count
         batchFailuresCount = 0
 
+        context.children.foreach { child =>
+          log.debug(s"unwatching and killing $child")
+          context.unwatch(child)
+          child ! PoisonPill
+        }
+
         // Restart syncing from the valid block available in state.
-        syncState = syncState.copy(
-          bestBlockHeaderNumber = firstCommonBlockNumber,
-          nextBlockToFullyValidate = firstCommonBlockNumber + 1
+        log.debug("starting with fresh SyncingHandler")
+        val syncingHandler = new SyncingHandler(
+          syncState.copy(
+            bestBlockHeaderNumber = firstCommonBlockNumber,
+            nextBlockToFullyValidate = firstCommonBlockNumber + 1,
+            pivotBlockUpdateFailures = 0
+          )
         )
-        masterPeer = Some(newMasterPeer)
```

Comment:
masterPeer gets updated elsewhere now?

Reply:
Hmm actually no; it is selected again after branch resolution. I'll change this to be closer to the (old) intended behavior. 👍